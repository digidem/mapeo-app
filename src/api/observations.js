// @flow
import fetch from 'fetch';
import { jsonRequest } from './base';

export interface CreateRequest {
  lat: number;
  lon: number;
}

export type UpdateRequest = {
  id: string
};

class Observation {
  static list = () =>
    jsonRequest({
      method: 'GET',
      route: '/observations'
    });

  static create = (observation: CreateRequest) =>
    jsonRequest({
      method: 'POST',
      route: '/observations',
      body: {
        device_id: '1',
        ...observation
      }
    });

  static update = (observation: UpdateRequest) =>
    jsonRequest({
      method: 'PUT',
      route: `/observations/${observation.id}`,
      body: {
        ...observation
      }
    });
}

export default Observation;
