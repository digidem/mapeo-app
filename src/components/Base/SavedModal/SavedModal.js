// @flow
import React from 'react';
import {
  Dimensions,
  View,
  Modal,
  Text,
  ImageBackground,
  StyleSheet
} from 'react-native';
import Image from 'react-native-remote-svg';
import type { Observation } from '../../../types/observation';
import type { Category } from '../../../types/category';
import Toast from '../Toast/Toast';
import I18n from 'react-native-i18n';
import moment from '../../../lib/localizedMoment';
import {
  WHITE,
  MAPEO_BLUE,
  LIGHT_GREY,
  VERY_LIGHT_BLUE,
  MEDIUM_GREY
} from '../../../lib/styles';
import getGPSText from '../../../lib/getGPSText';
import CategoryPin from '../../../images/category-pin.png';

export type StateProps = {
  selectedObservation: Observation,
  categories: {
    [id: string]: Category
  },
  gpsFormat: string,
  icons: Object
};

export type DispatchProps = {
  onHide: Function
};

const styles = StyleSheet.create({
  categoryContainer: {
    flex: 3,
    backgroundColor: VERY_LIGHT_BLUE,
    alignSelf: 'stretch',
    alignItems: 'center',
    borderBottomLeftRadius: 20,
    borderBottomRightRadius: 20
  },
  categoryPin: {
    width: 80,
    height: 90,
    marginTop: 15,
    justifyContent: 'center',
    alignItems: 'center'
  },
  confirmationModal: {
    alignSelf: 'center',
    alignItems: 'center',
    width: Dimensions.get('window').width * 0.8,
    height: Dimensions.get('window').height * 0.5,
    backgroundColor: 'white',
    borderRadius: 20
  },
  date: {
    color: MEDIUM_GREY,
    fontSize: 12,
    fontWeight: '400'
  },
  positionAtText: {
    fontSize: 12,
    color: 'black',
    fontWeight: '400'
  },
  positionText: {
    fontSize: 12,
    color: 'black',
    fontWeight: '700'
  },
  syncedContainer: {
    borderColor: LIGHT_GREY,
    borderBottomWidth: 1,
    flex: 1,
    alignSelf: 'stretch',
    justifyContent: 'center'
  },
  title: {
    fontSize: 20,
    fontWeight: '700',
    color: 'black',
    textAlign: 'center'
  }
});

I18n.fallbacks = true;
I18n.translations = {
  en: require('../../../translations/en'),
  es: require('../../../translations/es')
};

class SavedModal extends React.PureComponent<StateProps & DispatchProps> {
  render() {
    const {
      onHide,
      selectedObservation,
      categories,
      gpsFormat,
      icons
    } = this.props;
    const positionText = getGPSText({
      gpsFormat,
      lat: selectedObservation.lat,
      lon: selectedObservation.lon
    });
    let category;

    if (!selectedObservation) {
      return null;
    }

    if (selectedObservation.categoryId !== undefined) {
      category = categories[selectedObservation.categoryId];
    }

    return (
      <Toast onHide={onHide}>
        {selectedObservation && (
          <View
            style={{
              backgroundColor: 'rgba(52, 52, 52, 0.8)',
              flex: 1,
              alignItems: 'center',
              justifyContent: 'center'
            }}
          >
            <View style={styles.confirmationModal}>
              <View style={styles.syncedContainer}>
                <Text
                  style={{
                    textAlign: 'center',
                    fontSize: 20,
                    fontWeight: '700',
                    color: 'black'
                  }}
                >
                  {I18n.t('saved')}
                </Text>
              </View>
              <View style={styles.categoryContainer}>
                <ImageBackground
                  source={CategoryPin}
                  style={styles.categoryPin}
                >
                  {category && (
                    <View style={{ marginTop: -10 }}>
                      {!!category.icon &&
                        !!icons[category.icon] && (
                          <Image
                            source={{
                              uri: `data:image/svg+xml;utf8,${
                                icons[category.icon]
                              }`
                            }}
                            style={{ height: 30, width: 30 }}
                          />
                        )}
                    </View>
                  )}
                </ImageBackground>
                <View
                  style={{
                    justifyContent: 'center',
                    alignItems: 'center',
                    paddingBottom: 10
                  }}
                >
                  <Text style={styles.title}>
                    {selectedObservation.categoryId}
                  </Text>
                  <View style={{ flexDirection: 'row' }}>
                    <Text style={styles.positionAtText}>{I18n.t('at')} </Text>
                    <Text style={styles.positionText}>
                      {`${positionText}.`}
                    </Text>
                  </View>
                  <Text style={styles.date}>
                    {I18n.t('on')}{' '}
                    {I18n.currentLocale().indexOf('en') !== -1
                      ? moment(selectedObservation.created).format(
                          'MMMM D YYYY, h:hh A'
                        )
                      : moment(selectedObservation.created).format('LLL')}
                  </Text>
                </View>
              </View>
            </View>
          </View>
        )}
      </Toast>
    );
  }
}

export default SavedModal;
