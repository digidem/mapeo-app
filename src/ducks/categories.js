// @flow

import update from 'immutability-helper';
import { keyBy } from 'lodash';
import { create } from '../lib/redux';

export const {
  type: CATEGORY_LIST,
  action: categoryList,
  reducer: categoryListReducer
} = create('CATEGORY_LIST', {
  success: (state, meta, payload) => {
    const newState = update(state, {
      categories: { $set: keyBy(payload, 'id') }
    });

    return newState;
  }
});

export default [categoryListReducer];
