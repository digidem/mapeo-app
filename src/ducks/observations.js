// @flow

import update from 'immutability-helper';
import {keyBy} from 'lodash';
import {create} from '../lib/redux';
import {createSelector} from '../lib/selector';
import type {StoreState} from '../types/redux';
import type {Observation} from '../types/observation';

export const {
  type: OBSERVATION_LIST,
  action: observationList,
  reducer: observationListReducer
} = create('OBSERVATION_LIST', {
  success: (state, action) => {
    const observations = keyBy(action.payload, 'id');

    if (state.observations) {
      Object.keys(state.observations).forEach(id => observations[id] = {
        ...observations[id],
        ...state.observations[id],
      });
    }

    const newState = update(state, {observations: {$set: observations}});

    return newState;
  }
});

export const {
  type: OBSERVATION_CREATE,
  action: observationCreate,
  reducer: observationCreateReducer
} = create('OBSERVATION_CREATE', {
  start: (state, action) => {
    const observation = action.meta;

    if (!observation.lat && !observation.lon && state.gps.data) {
      observation.lat = state.gps.data.latitude;
      observation.lon = state.gps.data.longitude;
    }

    const newState = update(state, {selectedObservation: {$set: observation}});

    return newState;
  },
  success: (state, action) => {
    const newState = update(state, {selectObservation: {$set: action.payload}});

    return newState;
  }
});

export const {
  type: OBSERVATION_UPDATE,
  action: observationUpdate,
  reducer: observationUpdateReducer
} = create('OBSERVATION_UPDATE', {
  start: (state, action) => {
    let newState;
    if (state.selectedObservation &&
        action.meta.id === state.selectedObservation.id) {
      newState = update(state, {
        selectedObservation: {
          $set: {...state.selectedObservation, ...action.meta},
        }
      });
    } else {
      newState = update(state, {
        observations: {
          [action.meta.id]: {
            $set: {...state.observations[action.meta.id], ...action.meta}
          }
        }
      });
    }

    return newState;
  }
});

export const {
  type: OBSERVATION_SELECT,
  action: observationSelect,
  reducer: observationSelectReducer
} = create('OBSERVATION_SELECT', {
  start: (state, action) => {
    if (action.meta === undefined && state.selectedObservation) {
      return update(state, {
        selectedObservation: {$set: action.meta},
        observations: {
          [state.selectedObservation.id]: {
            $set: {
              ...(state.observations[state.selectedObservation.id] || {}),
              ...state.selectedObservation,
            }
          }
        }
      });
    } else {
      return update(state, {selectedObservation: {$set: action.meta}});
    }
  }
});

export const selectObservation = createSelector(
    [(state: StoreState, id: string): Observation =>
         state.app.observations[id]],
    (observation: Observation): Observation => observation,
    (observation: Observation, id: string): string => id);

export default [
  observationListReducer, observationCreateReducer, observationUpdateReducer,
  observationSelectReducer
];
