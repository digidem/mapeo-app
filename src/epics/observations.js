// @flow
import React from 'react';
import { Image } from 'react-native';
import type { ActionsObservable } from 'redux-observable';
import {
  OBSERVATION_LIST,
  observationList,
  OBSERVATION_CREATE,
  observationCreate,
  OBSERVATION_UPDATE,
  observationUpdate
} from '../ducks/observations';
import { Action } from '../types/redux';
import type { Observation as ObservationType } from '../types/observation';
import type { StoreState } from '../types/redux';
import Observation from '../api/observations';
import type { CreateRequest, UpdateRequest } from '../api/observations';

export const observationListEpic = (
  action$: ActionsObservable<Action<string, ObservationType[]>>,
  store: StoreState
) =>
  action$
    .ofType(OBSERVATION_LIST)
    .filter(action => action.status === 'Start')
    .flatMap(() =>
      Observation.list().map(observations => {
        console.log('RN - Observations List', observations);
        return observationList('', observations);
      })
    );

export const observationCreateEpic = (
  action$: ActionsObservable<Action<CreateRequest, ObservationType>>,
  store: StoreState
) =>
  action$
    .ofType(OBSERVATION_CREATE)
    .filter(action => action.status === 'Start')
    .flatMap(action =>
      Observation.create(action.meta).map(observation => {
        console.log('RN - Observation Created');
        return observationCreate(action.meta, observation);
      })
    );

// export const observationUpdateEpic = (
//   action$: ActionsObservable<Action<UpdateRequest, ObservationType>>,
//   store: StoreState
// ) =>
//   action$
//     .ofType(OBSERVATION_UPDATE)
//     .filter(action => action.status === 'Start')
//     .flatMap(action =>
//       Observation.update(action.meta).map(observation => {
//         console.log('RN - Observation Updated - ', action.meta, observation);
//         return observationUpdate(action.meta, observation);
//       })
//     );

export default [
  observationListEpic,
  observationCreateEpic
  // observationUpdateEpic
];
