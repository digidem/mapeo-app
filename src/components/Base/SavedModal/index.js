// @flow
import { connect } from 'react-redux';

import SavedModal from './SavedModal';
import { modalHide } from '../../../ducks/modals';
import { observationSelect } from '../../../ducks/observations';
import { observationSource } from '../../../ducks/observationSource';

const mapStateToProps = state => ({
  selectedObservation: state.app.selectedObservation,
  categories: state.app.categories,
  gpsFormat: state.app.settings.gpsFormat,
  icons: state.app.icons
});

const mapDispatchToProps = dispatch => ({
  onHide: () => {
    dispatch(modalHide('saved'));
    dispatch(observationSelect(undefined));
    dispatch(observationSource(undefined));
  }
});

export default connect(mapStateToProps, mapDispatchToProps)(SavedModal);
