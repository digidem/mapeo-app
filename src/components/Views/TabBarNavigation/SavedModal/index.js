// @flow
import { connect } from 'react-redux';
import { withNavigationFocus } from 'react-navigation';

import SavedModal from './SavedModal';
import { observationSelect } from '../../../../ducks/observations';
import { modalHide } from '../../../../ducks/modals';
import { observationSelect } from '../../../../ducks/observations';

const mapStateToProps = state => ({
  selectedObservation: state.app.selectedObservation
});

const mapDispatchToProps = dispatch => ({
  onHide: () => {
    dispatch(modalHide('saved'));
    dispatch(observationSelect(undefined));
  }
});

export default connect(mapStateToProps, mapDispatchToProps)(SavedModal);
