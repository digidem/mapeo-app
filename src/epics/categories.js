// @flow
import React from 'react';
import type { ActionsObservable } from 'redux-observable';
import { Image } from 'react-native';
import { CATEGORY_LIST, categoryList } from '../ducks/categories';
import { PRESETS_SELECT, presetsSelect } from '../ducks/fields';
import { createCategory } from '../mocks/categories';
import type { Action } from '../types/redux';
import type { Category } from '../types/category';
import Presets from '../api/presets';

export const categoryListEpic = (
  action$: ActionsObservable<Action<string, Category[]>>
) =>
  action$
    .ofType(CATEGORY_LIST)
    .filter(action => action.status === 'Start')
    .flatMap(() => Presets.list().map(list => presetsSelect(list[0])));

export default [categoryListEpic];
