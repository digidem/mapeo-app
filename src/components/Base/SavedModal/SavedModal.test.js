// @flow
import React from 'react';
import 'react-native';
import renderer from 'react-test-renderer';
import SavedModal from './SavedModal';
import { createObservation } from '../../../mocks/observations';
import { createCategory } from '../../../mocks/categories';

describe('SavedModal tests', () => {
  test('snapshots', () => {
    const observation = createObservation();
    const categories = {
      [observation.categoryId]: createCategory()
    };
    const icons = {
      [createCategory().icon]: 'icon'
    };
    const onHide = jest.fn();

    const props = [{ observation }];

    let tree;
    props.forEach(p => {
      tree = renderer
        .create(
          <SavedModal
            selectedObservation={p.observation}
            onHide={onHide}
            categories={categories}
            icons={icons}
            gpsFormat="utm"
          />
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
