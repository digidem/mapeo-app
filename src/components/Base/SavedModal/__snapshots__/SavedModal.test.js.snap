// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SavedModal tests snapshots 1`] = `
<Modal
  animation="slide"
  hardwareAccelerated={false}
  onRequestClose={[Function]}
  transparent={true}
  visible={true}
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(52, 52, 52, 0.8)",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "center",
          "backgroundColor": "white",
          "borderRadius": 20,
          "height": 667,
          "width": 600,
        }
      }
    >
      <View
        style={
          Object {
            "alignSelf": "stretch",
            "borderBottomWidth": 1,
            "borderColor": "#EAEAEA",
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "black",
              "fontSize": 20,
              "fontWeight": "700",
              "textAlign": "center",
            }
          }
        >
          Saved!
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "alignSelf": "stretch",
            "backgroundColor": "#ccffff",
            "borderBottomLeftRadius": 20,
            "borderBottomRightRadius": 20,
            "flex": 3,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "height": 90,
              "justifyContent": "center",
              "marginTop": 15,
              "width": 80,
            }
          }
        >
          <Image
            source={
              Object {
                "testUri": "../../../src/images/category-pin.png",
              }
            }
            style={
              Array [
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
                Object {
                  "height": 90,
                  "width": 80,
                },
                undefined,
              ]
            }
          />
          <View
            style={
              Object {
                "marginTop": -10,
              }
            }
          >
            <View
              pointerEvents="none"
              style={
                Array [
                  Object {
                    "height": 30,
                    "width": 30,
                  },
                  undefined,
                ]
              }
            >
              <View
                style={
                  Object {
                    "flex": 1,
                  }
                }
              >
                <RCTWebView
                  allowsInlineMediaPlayback={undefined}
                  automaticallyAdjustContentInsets={undefined}
                  bounces={undefined}
                  contentInset={undefined}
                  dataDetectorTypes={undefined}
                  decelerationRate={undefined}
                  injectedJavaScript={undefined}
                  mediaPlaybackRequiresUserAction={undefined}
                  messagingEnabled={false}
                  onLoadingError={[Function]}
                  onLoadingFinish={[Function]}
                  onLoadingStart={[Function]}
                  onMessage={[Function]}
                  onShouldStartLoadWithRequest={undefined}
                  scalesPageToFit={true}
                  scrollEnabled={false}
                  source={
                    Object {
                      "html": "<html><head><style>html, body { margin:0; padding:0; overflow:hidden; background-color: transparent; } svg { position:fixed; top:0; left:0; height:100%; width:100% }</style></head><body>n</body></html>",
                    }
                  }
                  style={
                    Array [
                      Object {
                        "flex": 1,
                      },
                      Object {
                        "backgroundColor": "#ffffff",
                      },
                      Array [
                        Object {
                          "backgroundColor": "transparent",
                          "height": 100,
                          "width": 200,
                        },
                        Object {
                          "height": 30,
                          "width": 30,
                        },
                      ],
                    ]
                  }
                />
              </View>
            </View>
          </View>
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "justifyContent": "center",
              "paddingBottom": 10,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "black",
                "fontSize": 20,
                "fontWeight": "700",
                "textAlign": "center",
              }
            }
          >
            19
          </Text>
          <View
            style={
              Object {
                "flexDirection": "row",
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "black",
                  "fontSize": 12,
                  "fontWeight": "400",
                }
              }
            >
              at
               
            </Text>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "black",
                  "fontSize": 12,
                  "fontWeight": "700",
                }
              }
            >
              50.5, 50.5.
            </Text>
          </View>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#666666",
                "fontSize": 12,
                "fontWeight": "400",
              }
            }
          >
            on
             
            February 25 2018, 12:12 PM
          </Text>
        </View>
      </View>
    </View>
  </View>
</Modal>
`;
